### PR TITLE
Implement item_candidates for Discogs plugin

### PR DESCRIPTION
**Pull request by ghbrown**
## Description

Implements `item_candidates` for the Discogs plugin, which begins to address #2544 and #3574.
Continues staled PR #3682 by implementing a subset of its provided features as requested by @sampsyo.

**Major change from  #3682**: no longer using string distance to narrow matches before returning candidates. This additional narrowing seems reasonable but it slowed down `item_candidates` (though marginally compared to the API call) and more importantly used a string match tolerance that users had no obvious way to control as far as I could tell. It seems preferable for as much pruning/matching to happen in a single unified function which would take all candidates returned by `item_candidates` as input; hardcoded matching in a separate function could result in harder to control and debug behavior.

**Known fragility**: results can be quite sensitive to the Discogs query, which is of the form  of string `<artist> <track>`, for example a difference of a single word in only the title can result in no results. In general this behavior seems to be known and documented in a few places, but I figured I'd mention it.

```
Ameega Funkysize My Baby  // wrong title, no results
Ameega Funkysize Now Baby // correct artist and title, query finds proper album
```

Querying only `<artist>` then using the first few album results performed significantly worse empirically, so `<artist> <track>` was used as the query usually returns the exact album containing the track if all data is correct.

## To Do

I will address these To Dos, but would like some feedback on the first draft first if possible.
@JOJ0 If you can give any feedback that would be great! One thing I'm currently worried about is testing. I run `tox`, but it doesn't take that long (about 10 seconds), and many things seem to be skipping. My best guess is that my version of Python (3.9.9) is too new, so maybe I'll have to set up an environment with 3.8 and `beets`, but for now I'm posting the code it can be discussed and tested via CI.

- [x] Documentation.
- [x] Changelog.
- [x] Tests.

### CODE REVIEW

File: beetsplug/discogs.py
Comment (JOJ0) [inline beetsplug/discogs.py:197]: Around here you could do an early exit, similar to that one in `candidates`: https://github.com/ghbrown/beets/blob/discogs_s_ic/beetsplug/discogs.py#L161-L163
It would address your legimite concerns described in "Known fragility"

Back then, when I submitted this early exit in `candidates` I was not sure if such an exit should also happen when _only_ artist is given. As you mentionend, results are pretty bad and almost useless then. But the good thing is, in a verbose output a user would see something like this, which tells that album is missing and only artist is considered in the search: `Search terms: Dave Clarke - Remixes & Rarities 1992-2005` (happening in `beets.autotag.match.py`)

Since something similar is coming from that module: `Item search terms: <artist> <title>`, I think an early exit if both artist and title are missing it is sufficient and most of all streamlines the behaviour with regular album discogs search.

**Review by JOJ0 — APPROVED** [review]
About running tests locally I'm not an expert either but these helped me in the past:

Lint with `tox -e lint`

Run the specific test for the discogs plugin: `python -m unittest discover -p 'test_discogs*'`

I even started to play around with the new feature in my test environment: I do get results but one thing that I'm missing is the output in the candidates list telling me that I'm actually looking at a Discogs result. In album view this looks something like this:

```
Candidates:
1. Dave Clarke - Remixes & Rarities 1992-2005 (77.6%) (id, tracks) (Discogs, Vinyl, 2006, Belgium, Music Man Records, MM129C)
...
```

In singleton mode the output of `(Discogs, ...` is missing:
```
Candidates:
1. Jark Prongo - Movin Thru Your System (Dave Clarke Remix) (52.6%) (id, index, length)
...
```

File: beetsplug/discogs.py
Comment (ghbrown) [reply beetsplug/discogs.py:197]: Thanks for the suggestion. I agree, I think early exiting only when both track name and artist are missing is a better choice for now.

**Comment by ghbrown** [comment]
> one thing that I'm missing is the output in the candidates list telling me that I'm actually looking at a Discogs result

> In singleton mode the output of (Discogs, ... is missing

Thanks for giving it a shot and for this note. It took me too long to figure out why it was happening, so I'll write it up just in case it's useful to you.

- Since the Discogs plugin doesn't yet inherit `MetadataSourcePlugin` (as you warned me about) one can't provide `data_source` to returned objects in such a global way (like the Spotify plugin does, for example). So for albums the Discogs plugin manually adds `data_source` to the `AlbumInfo` object [here](https://github.com/beetbox/beets/blob/6989bce6cab134e937acd21720b4a165f916a60b/beetsplug/discogs.py#L364). I've essentially had to do the same by looping through the candidates returned by `item_candidates` and setting the field, which works fine.

On a related note, the `data_source` is only added to the disambiguation string if it isn't `MusicBrainz`. This is weird and probably related to the original source being Musicbrainz. This behavior is hardcoded in [`ui/command.py`](https://github.com/beetbox/beets/blob/6989bce6cab134e937acd21720b4a165f916a60b/beets/ui/commands.py#L193). If this is going to be changed (I think it should be in another PR) my vote is to only print data sources (and all of them, even MusicBrainz) only when there are multiple data sources. So the only time that data source wouldn't be provided is when there is only a single source (which could be MusicBrainz, Discogs, Spotify, or anything). I'm not sure what the sentiment on "MusicBrainz-special-casing" is, but I can file an issue to discuss this with the community if that sounds reasonable.

**Comment by ghbrown** [comment]
My main complaint now is that it sometimes returns far too many candidates. I've seen as many as 70, which is more than enough to inconveniently push the most promising results off the top of the terminal.

What's the best way to control this, or does it not need controlling? The previous implementation from the staled PR used `string_dist` internally to limit the number of returned candidates, but I have some reservations about this as I discussed in my original comment.

**Comment by JOJ0** [comment]
I just tested a little, even on my production library :-o Everything went pretty fine. I imported some older single drumandbass tracks and I have to say: I love it! Thank you already for a great and long needed feature, at least on my end. It makes my life so much easier to finally have a way to search for track names (and not album names) from Discogs. MusicBrainz is not very complete when it comes to DJ/electronic music, you know, so MB singleton tagging was never an option for my many single files.....and album tagging single files is tedious as well since album searches do not consider track name tags.

Anyway: Because of that many responses: I also realised that it's quite a lot. The most important thing though is that more often than not I find the correct ones in the top 5. I can't help much at the moment with an idea of how to reduce it. You might want to have a look at the MusicBrainz autotag implementation and also how other metadata source plugins do it. A few code links that might help....or not... :-)

https://github.com/beetbox/beets/blob/master/beets/autotag/hooks.py#L635-L647

https://github.com/beetbox/beets/blob/master/beets/autotag/mb.py#L552-L571

MusicBrainz settings offer to limit the search results total, at least I'm aware of it for album search, not sure about singletons but I think it is a general limit that is sent to the MB API, telling that not more than this number should be returnend - as simple as that. The default is 5.

Also, now going a little off-topic, since I faced the issue of old and wrong album info on those original tracks (album metadatafield simply containend the track names for some reason !?!) I want to mention this comment here and was thinking if a new issue should be openend for discussing this question: https://github.com/beetbox/beets/blob/master/beets/autotag/__init__.py#L94-L95
In my opinion, if this was to be implemented, the user should have the freedom of configuring whether album field should be emptied or not!

**Comment by ghbrown** [comment]
I'm glad to hear it's working out for you!

> I want to mention this comment here and was thinking if a new issue should be openend for discussing this question

This seems like a broader scoped issue than the Discogs plugin, right? I'm not sure I exactly understand your problem, but I remember hearing that plugins like zeros (I think that was the name) could clear/null metadata for you. If that doesn't cover it I would say a new issue is the way to go.

> Either you do some extra work and make this configurable for the singletons search already or you find a reasonable hardcoded value (10 to be safe?) for now and open an issue/feature reuqest for implementing cvonfigurable search results, for both album and singleton searches on Discogs.

I'm going to go with the latter, since those changes also feel a bit out of scope on this PR. I think 10 is a good number, and since the Discogs API search should already internally rank results by relevance we're hopefully not losing much.

**Comment by ghbrown** [comment]
I'll fix that lint error when I add tests, docs, etc., if there are no more changes to the code itself.

**Comment by JOJ0** [comment]
I merged this feature into my "production beets version" and use it on a daily basis. Works very well for my needs. I think it doesn't hurt to finish this PR with release notes and docs, so all checkmarks are ticked :-) If maintainers request final changes you could still add them.

**Comment by JOJ0** [comment]
I just realised one thing that might be missing: Other metadata sources show the URL of the found track:

```
/remote/data/music-archive/DrumnbassJungleDrumfunk/0-Single-Tracks/Crossfire/Crossfire - Firebolt.mp3
Correcting track tags from:
     - Crossfire - Firebolt
To:
    Neira, Miguel Aimeur, DJ CÃ³las NG - Firebolt
URL:
    https://beatport.com/track/firebolt/15316747
(Similarity: 71.6%) (title, source) (Beatport)
[A]pply, More candidates, Skip, Use as-is, Enter search, enter Id, aBort,
eDit, edit Candidates, plaY? m
```

```
Correcting track tags from:
    crossfire  -  Battlefield
To:
    Crossfire - Hysterical Rochords
URL:
    https://musicbrainz.org/recording/64a2e940-0158-4eb0-9b63-1d32075aa69e
(Similarity: 53.3%) (title)
Apply, More candidates, Skip, Use as-is, Enter search, enter Id, aBort,
eDit, edit Candidates, plaY? m
```

Not sure if that would be easily possible with Discogs. It would be the URL of the release, since there is no such thing as an recording/track URL.

File: beetsplug/discogs.py
Comment (JOJ0) [inline beetsplug/discogs.py:235]: I played around a little and found out that simply applying data_url from the album here would do the trick:
```
            track_info['data_url'] = album_info.data_url
```

I think the `if not` is not required since data_url will probably always be required.

Looks like that then :-)

```
/remote/data/music-library/DrumnbassJungleDrumfunk/0-Album-Missing/Crossfire - Firebolt.mp3
Tagging track: Crossfire - Firebolt
URL:
    https://www.discogs.com/release/125946-Crossfire-Counterforce-Firebolt
(Similarity: 93.3%) (source) (Discogs)
[A]pply, More candidates, Skip, Use as-is, Enter search, enter Id, aBort,
eDit, edit Candidates, plaY?
```

Comment (ghbrown) [reply beetsplug/discogs.py:235]: Thanks for digging into this! I'll make that edit.

I don't think that `if not` hurts anything (besides performance negligibly). The real question for me is: if there is a URL there already would it be something reasonable and would we want to pass it along, or is the URL we're providing via `album_info.data_url` always the "right one"?
For me the answer to this question would determine whether to use the `if not`.

**Comment by ghbrown** [comment]
Sounds good, I'll get started on the other unchecked boxes.

**Comment by ghbrown** [comment]
As a summary for the latest commit(s):
- a helper function for `item_candidates()` used to parse the returned albums into their respective `TrackInfo`s, adding artist and artist ID data to the tracks that were concatenated into a (later returned) `candidates` list, but this feature was almost entirely redundant with the `.tracks` fields of `AlbumInfo`s and actually provided some features that should arguable have been present in `get_album_info()`
- the most recent commit (as of writing) removes the helper function in favor of adding these fields to all tracks of an album in `get_album_info()` in a way that improves the robustness of `get_album_info()`

The root of this whole issue seems to be that `get_track_info()` never actually finds a usable track artist and always returns an empty string (in my various tests). Now we detect case this in `get_album_info()` and supply the album artist to each track as a stand-in before returning the `AlbumInfo` object.

Whether we supply the `data_url` and `data_source` field to each `TrackInfo` in an object is probably more a matter of taste and may mean some kind of helper function needs to re-enter the picture to specifically provide this info only when matching singletons.

File: beetsplug/discogs.py
Comment (JOJ0) [reply beetsplug/discogs.py]: I'm pretty sure album_info.data_url is the _only_ one we'll get, haha ;-) Certainly you are right, the `if not` doesn't really hurt. I just though that it would even be cleaner and "kind of a statement" to say here: There is no track_url, because it's just not existing on discogs, thus we simply apply the album_url to the track so the importer prompt shows it similar to the track_urls musicbrainz and other metadata source plugins do it.

Here some debugging around these lines: FIXME

```
$ beet import -I -t -s /remote/data/music-archive/DrumnbassJungleDrumfunk/0-Single-Tracks/Crossfire/Crossfire\ -\ Hideout-sour.mp3 

This is data_url in get_album_info:

https://www.discogs.com/release/76357-Crossfire-Hideout-Swordfight



An this is the whole result.data object in get_album_info:

{'artists': [{'anv': '',
              'id': 3701,
              'join': '',
              'name': 'Crossfire',
              'resource_url': 'https://api.discogs.com/artists/3701',
              'role': '',
              'thumbnail_url': 'https://i.discogs.com/F7Z4BCbnKCkFCls871e6X5DnxO7zpnUX760MlkmKi7Q/rs:fit/g:sm/q:40/h:222/w:600/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9BLTM3MDEt/MTQ4NzI1MzU5Ni0z/NTcwLmpwZWc.jpeg',
              'tracks': ''}],
 'artists_sort': 'Crossfire',
 'barcode': ['5 023903 216700',
             '1210 006 A.',
             '1210 006 AA. NEIL MASTERPIECE '],
 'blocked_from_sale': False,
 'catno': '1210-006',
 'community': {'contributors': [{'resource_url': 'https://api.discogs.com/users/milo',
                                 'username': 'milo'},
                                {'resource_url': 'https://api.discogs.com/users/The_Undertaker',
                                 'username': 'The_Undertaker'},
                                {'resource_url': 'https://api.discogs.com/users/Dandruff',
                                 'username': 'Dandruff'},
                                {'resource_url': 'https://api.discogs.com/users/Aphex139',
                                 'username': 'Aphex139'},
                                {'resource_url': 'https://api.discogs.com/users/the_electrician',
                                 'username': 'the_electrician'},
                                {'resource_url': 'https://api.discogs.com/users/Zubeldia',
                                 'username': 'Zubeldia'},
                                {'resource_url': 'https://api.discogs.com/users/pano9000',
                                 'username': 'pano9000'},
                                {'resource_url': 'https://api.discogs.com/users/jmills',
                                 'username': 'jmills'},
                                {'resource_url': 'https://api.discogs.com/users/mhoys',
                                 'username': 'mhoys'},
                                {'resource_url': 'https://api.discogs.com/users/OllieDee',
                                 'username': 'OllieDee'},
                                {'resource_url': 'https://api.discogs.com/users/vinylmaddy',
                                 'username': 'vinylmaddy'}],
               'data_quality': 'Needs Vote',
               'have': 360,
               'rating': {'average': 3.86, 'count': 66},
               'status': 'Accepted',
               'submitter': {'resource_url': 'https://api.discogs.com/users/milo',
                             'username': 'milo'},
               'want': 80},
 'companies': [{'catno': '',
                'entity_type': '30',
                'entity_type_name': 'Lacquer Cut At',
                'id': 120735,
                'name': 'Masterpiece',
                'resource_url': 'https://api.discogs.com/labels/120735',
                'thumbnail_url': 'https://i.discogs.com/D4k3gvHSqgSNPVXGJ96T5_4O5Y5ep_HQ7TsGpDAWxOI/rs:fit/g:sm/q:40/h:600/w:600/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9MLTEyMDcz/NS0xNTI2MzEyODQ3/LTg4ODQuanBlZw.jpeg'},
               {'catno': '',
                'entity_type': '9',
                'entity_type_name': 'Distributed By',
                'id': 306370,
                'name': 'Vinyl UK',
                'resource_url': 'https://api.discogs.com/labels/306370',
                'thumbnail_url': 'https://i.discogs.com/oKWJ70-YpIGeM8_24PvCJuHyTdRlW5vPxoMvnNcGMAc/rs:fit/g:sm/q:40/h:150/w:150/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9MLTMwNjM3/MC0xNTU1MzUyMTI5/LTQ4ODkuanBlZw.jpeg'},
               {'catno': '',
                'entity_type': '21',
                'entity_type_name': 'Published By',
                'id': 48207,
                'name': 'Copyright Control',
                'resource_url': 'https://api.discogs.com/labels/48207',
                'thumbnail_url': 'https://i.discogs.com/FRvFvNsk726fp6xh-pQzN18g_vPp3VwBfgjJEZY3SSg/rs:fit/g:sm/q:40/h:150/w:150/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9MLTQ4MjA3/LTEzNTY5MDQ0ODct/MTkzNy5qcGVn.jpeg'}],
 'country': 'UK',
 'cover_image': 'https://i.discogs.com/p0m8SdTvQcRL7PdbbYxLsqhhG8SM0w1DyQPFn0u9VRg/rs:fit/g:sm/q:90/h:406/w:400/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9SLTc2MzU3/LTExNDE5Mzk3Njgu/anBlZw.jpeg',
 'data_quality': 'Needs Vote',
 'date_added': '2002-11-28T07:40:02-08:00',
 'date_changed': '2022-04-27T21:48:57-07:00',
 'estimated_weight': 230,
 'extraartists': [{'anv': 'Neil',
                   'id': 342597,
                   'join': '',
                   'name': 'Neil Devine',
                   'resource_url': 'https://api.discogs.com/artists/342597',
                   'role': 'Lacquer Cut By',
                   'thumbnail_url': 'https://i.discogs.com/8ffMnvAUQrUFEKPvI77cKuH_OMlz8on4XDp30xtmCqI/rs:fit/g:sm/q:40/h:200/w:200/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9BLTM0MjU5/Ny0xNTI2MzEyMzA5/LTQzMDMuanBlZw.jpeg',
                   'tracks': ''},
                  {'anv': 'D. Van Simaeys',
                   'id': 692992,
                   'join': '',
                   'name': 'Dimitry Van Simaeys',
                   'resource_url': 'https://api.discogs.com/artists/692992',
                   'role': 'Written-By, Producer',
                   'thumbnail_url': '',
                   'tracks': ''}],
 'format': ['Vinyl', '12"', '45 RPM'],
 'format_quantity': 1,
 'formats': [{'descriptions': ['12"', '45 RPM'], 'name': 'Vinyl', 'qty': '1'}],
 'genre': ['Electronic'],
 'genres': ['Electronic'],
 'id': 76357,
 'identifiers': [{'description': 'Sleeve',
                  'type': 'Barcode',
                  'value': '5 023903 216700'},
                 {'description': 'Side A',
                  'type': 'Matrix / Runout',
                  'value': '1210 006 A.'},
                 {'description': 'Side B',
                  'type': 'Matrix / Runout',
                  'value': '1210 006 AA. NEIL MASTERPIECE '}],
 'images': [{'height': 406,
             'resource_url': 'https://i.discogs.com/p0m8SdTvQcRL7PdbbYxLsqhhG8SM0w1DyQPFn0u9VRg/rs:fit/g:sm/q:90/h:406/w:400/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9SLTc2MzU3/LTExNDE5Mzk3Njgu/anBlZw.jpeg',
             'type': 'primary',
             'uri': 'https://i.discogs.com/p0m8SdTvQcRL7PdbbYxLsqhhG8SM0w1DyQPFn0u9VRg/rs:fit/g:sm/q:90/h:406/w:400/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9SLTc2MzU3/LTExNDE5Mzk3Njgu/anBlZw.jpeg',
             'uri150': 'https://i.discogs.com/lovwaNWoNo7n5_dg7hmp4zCT6PsHFkxltPQm-ChmQgM/rs:fit/g:sm/q:40/h:150/w:150/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9SLTc2MzU3/LTExNDE5Mzk3Njgu/anBlZw.jpeg',
             'width': 400},
            {'height': 406,
             'resource_url': 'https://i.discogs.com/5RhBGHon8hAhOHOJRjWP-ZpXAM0l5BTYIOFMhAeXWR8/rs:fit/g:sm/q:90/h:406/w:400/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9SLTc2MzU3/LTExNDE5Mzk3Nzku/anBlZw.jpeg',
             'type': 'secondary',
             'uri': 'https://i.discogs.com/5RhBGHon8hAhOHOJRjWP-ZpXAM0l5BTYIOFMhAeXWR8/rs:fit/g:sm/q:90/h:406/w:400/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9SLTc2MzU3/LTExNDE5Mzk3Nzku/anBlZw.jpeg',
             'uri150': 'https://i.discogs.com/XV2hH6p-hwa3WkeuoNZT-Vv5sjQxybQumErK94TC9O8/rs:fit/g:sm/q:40/h:150/w:150/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9SLTc2MzU3/LTExNDE5Mzk3Nzku/anBlZw.jpeg',
             'width': 400},
            {'height': 598,
             'resource_url': 'https://i.discogs.com/LMSfQzLgS1vEmkvVG7mqzTrK6Dwply5HCTt1CMcY_bE/rs:fit/g:sm/q:90/h:598/w:600/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9SLTc2MzU3/LTE2MTQ5ODQ5ODEt/NTQ3Ny5qcGVn.jpeg',
             'type': 'secondary',
             'uri': 'https://i.discogs.com/LMSfQzLgS1vEmkvVG7mqzTrK6Dwply5HCTt1CMcY_bE/rs:fit/g:sm/q:90/h:598/w:600/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9SLTc2MzU3/LTE2MTQ5ODQ5ODEt/NTQ3Ny5qcGVn.jpeg',
             'uri150': 'https://i.discogs.com/6HyQ2zEmlCUTI3H658oDhEtAWYKzrOnSNUZhhnpzTNE/rs:fit/g:sm/q:40/h:150/w:150/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9SLTc2MzU3/LTE2MTQ5ODQ5ODEt/NTQ3Ny5qcGVn.jpeg',
             'width': 600},
            {'height': 592,
             'resource_url': 'https://i.discogs.com/g6_3YMeMA2f4bKNZCZDji_36bfJPtl8IgarxhZF72Wo/rs:fit/g:sm/q:90/h:592/w:600/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9SLTc2MzU3/LTE2MTQ5ODQ5OTQt/OTU3NS5qcGVn.jpeg',
             'type': 'secondary',
             'uri': 'https://i.discogs.com/g6_3YMeMA2f4bKNZCZDji_36bfJPtl8IgarxhZF72Wo/rs:fit/g:sm/q:90/h:592/w:600/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9SLTc2MzU3/LTE2MTQ5ODQ5OTQt/OTU3NS5qcGVn.jpeg',
             'uri150': 'https://i.discogs.com/wn5Ticd8KZa3wqj_8eAgjs4gPynxWEcw1MxWxXNdHnM/rs:fit/g:sm/q:40/h:150/w:150/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9SLTc2MzU3/LTE2MTQ5ODQ5OTQt/OTU3NS5qcGVn.jpeg',
             'width': 600}],
 'label': ['1210 Recordings', 'Masterpiece', 'Vinyl UK', 'Copyright Control'],
 'labels': [{'catno': '1210-006',
             'entity_type': '1',
             'entity_type_name': 'Label',
             'id': 182264,
             'name': '1210 Recordings',
             'resource_url': 'https://api.discogs.com/labels/182264',
             'thumbnail_url': 'https://i.discogs.com/nWDUE88YKF6q_BbbBAC1WppiNNvo2Nj_EQpDTyr0BfE/rs:fit/g:sm/q:40/h:418/w:422/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9MLTE4MjI2/NC0xMzcwOTg1MzMw/LTM5NjYuanBlZw.jpeg'}],
 'lowest_price': 2.17,
 'master_id': 225941,
 'master_url': 'https://api.discogs.com/masters/225941',
 'notes': 'Made in England',
 'num_for_sale': 15,
 'released': '2002',
 'released_formatted': '2002',
 'resource_url': 'https://api.discogs.com/releases/76357',
 'series': [],
 'status': 'Accepted',
 'style': ['Drum n Bass'],
 'styles': ['Drum n Bass'],
 'thumb': 'https://i.discogs.com/lovwaNWoNo7n5_dg7hmp4zCT6PsHFkxltPQm-ChmQgM/rs:fit/g:sm/q:40/h:150/w:150/czM6Ly9kaXNjb2dz/LWRhdGFiYXNlLWlt/YWdlcy9SLTc2MzU3/LTExNDE5Mzk3Njgu/anBlZw.jpeg',
 'title': 'Hideout / Swordfight',
 'tracklist': [{'duration': '6:48',
                'position': 'A',
                'title': 'Hideout',
                'type_': 'track'},
               {'duration': '6:22',
                'position': 'B',
                'title': 'Swordfight',
                'type_': 'track'}],
 'type': 'release',
 'uri': 'https://www.discogs.com/release/76357-Crossfire-Hideout-Swordfight',
 'user_data': {'in_collection': False, 'in_wantlist': False},
 'videos': [{'description': 'Official dj Crossfire page   »  '
                            'www.facebook.com/crossfirednb   »   '
                            'crossfire.bandcamp.com   »   '
                            'www.soundcloud.com/crossfire',
             'duration': 416,
             'embed': True,
             'title': 'Crossfire - Hideout (Original Mix)',
             'uri': 'https://www.youtube.com/watch?v=kIZll3BmDI4'},
            {'description': 'Official dj Crossfire page   »  '
                            'www.facebook.com/crossfirednb   »   '
                            'crossfire.bandcamp.com   »   '
                            'www.soundcloud.com/crossfire',
             'duration': 391,
             'embed': True,
             'title': 'Crossfire - Swordfight (Original Mix)',
             'uri': 'https://www.youtube.com/watch?v=mHc_i6AhIZk'}],
 'year': 2002}
```

Comment (JOJ0) [reply beetsplug/discogs.py]: Ah! Sorry! I should have read your 2 latest commits where you actually adress this and make sure data_url from album_info is applied to each track! Nice!

**Comment by JOJ0** [comment]
I like your decision of getting rid of the get_tracks_from_album method! Simplifies the code, easier to read, just nice!

**Comment by ghbrown** [comment]
Thanks! I got distracted looking at the existing tests and realized something was a little weird.

Speaking of tests, since I only implemented `item_candidates` I'm not sure if there are any Discogs-specific tests to write. It seems the `test_<plugin>.py` files only test things like whether the artist, etc. were parsed and written correctly. I've looked through these files for other plugins and I see nothing that tests `(item_)candidates` or anything like that. I suspect since this interface is common to a lot of source plugins this happens in `test_plugins.py` or somewhere like that, but either way my tentative conclusion is that I don't need to write any new tests(?).

**Comment by ghbrown** [comment]
I'm (optimistically) checking all the boxes, with testing and my changes to the Discogs plugin's `get_album_info()` being the biggest unknowns right now (reviewers see my previous two comments). Either way, I think this PR is quite near completion.

File: docs/plugins/discogs.rst
Comment (JOJ0) [inline docs/plugins/discogs.rst:103]: I suggest to remove the "in progress" state. Once this PR gets merged I think it will be a grown-up feature and should be stated as that in the docs :-) If new bugs get introduced we need to fix them later on but from my point of view, this feature will probably do it's job as intended.

Also the "is not working well" sounds a little negative. I still think it is very helpful that we mention it as a alternative way of importing files that are not in a album-related folder, so maybe just rephrase.

File: docs/changelog.rst
Comment (JOJ0) [inline docs/changelog.rst:40]: Dot  to finish statement but that's often forgotten by changelog writers, so no worries ;-)

**Review by JOJ0 — CHANGES_REQUESTED** [review]
As my first act as a newly joined member of the Beets project I'd like to merge this PR. Just some tiny changes and I think we are ready to merge.

File: docs/plugins/discogs.rst
Comment (ghbrown) [reply docs/plugins/discogs.rst:103]: I agree this PR provides mature functionality, but my understanding was that there are other features to add/interfaces to implement before singleton matching is complete. The original staled PR also implemented `track_for_id`, but Adrian asked that `item_candidates` be separated.

I say we list this limitation directly in place of the "in progress" line, something like "matching tracks by ID not yet supported". I'll rephrase the tip as well.

File: docs/changelog.rst
Comment (ghbrown) [reply docs/changelog.rst]: Added, thanks! I also fixed the line above.

**Comment by ghbrown** [comment]
@JOJ0 Congratulations on your position!

I addressed the changes you requested. Please let me know if I can do anything else.

File: docs/plugins/discogs.rst
Comment (JOJ0) [inline docs/plugins/discogs.rst:103]: Oh! I completely forgot the fact that `-s` would not work if one would try to supply a Discogs ID with `-S`. I didn't test it yet but what will happen? Ugly error or nothing?

Thanks for mentioning that it doesn't work yet. Very good!

**Review by JOJ0 — COMMENTED** [review]
Alright FIXME

**Review by JOJ0 — APPROVED** [review]

File: docs/plugins/discogs.rst
Comment (ghbrown) [reply docs/plugins/discogs.rst:103]: I get something like this if I try to use IDs

```
Enter recording ID: r3508566  # the ID for album containing song
No matching records found
[S]kip, Use as-is, ...
```

which is not insightful, but also perfectly reasonable. I think documentation that this feature isn't implemented is fine for now, I don't think changing the interface to display this fact would be worthwhile.

**Comment by ghbrown** [comment]
> Alright, what do you think about the `-s` for ID feature? Is it complicated to implement? You wanna leave it out in this PR and first get this one merged?

In the original staled PR Adrian asked for `item_candidates` to be separated and merged first, I believe because the original strategy for adding the ID functionality changed some important and frequently used classes.

That contributor may have had a good reason for making those changes (I haven't looked deeply at the ID feature yet), but I think it would probably be best to revisit their implementation and see if we can do it without such impactful changes in a separate PR.

**Comment by JOJ0** [comment]
Ok, I agree, let's put the searching by ID feature aside for another day and another PR. I will open an issue about this missing feature and merged this one :tada: Thanks so much for your work, this is a very handy addition to beets! If you feel like it and would like to continue work on the Discogs plugin, you're certainly very welcome anytime! :-)

**Pull request merged**
